In [9]:
#importera paket
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats as scs 

# import cars data
filepath = '../Uppgift/Data/auto-mpg.csv'
cars = pd.read_csv(filepath)
cars = cars.dropna() # Remove rows containing NaN

In [8]:
cars.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

#### what is the 95% confidence interval for average fuel effiency of cars with 4 cylinders?

In [16]:
from scipy.stats import t # Import the t-distribution from SciPy

cylinder4 = cars.query('cylinders == 4') # subset data for 4 cylinders cars
cylinder4.head()
 

mpg  cylinders  displacement  horsepower  weight  acceleration  \
14  24.0          4         113.0        95.0    2372          15.0   
18  27.0          4          97.0        88.0    2130          14.5   
19  26.0          4          97.0        46.0    1835          20.5   
20  25.0          4         110.0        87.0    2672          17.5   
21  24.0          4         107.0        90.0    2430          14.5   

    model_year  origin                          name  
14          70   japan         toyota corona mark ii  
18          70   japan                  datsun pl510  
19          70  europe  volkswagen 1131 deluxe sedan  
20          70  europe                   peugeot 504  
21          70  europe                   audi 100 ls

In [4]:
mean = np.mean(cylinder4['mpg']) # Calculate the sample mean
std = np.std(cylinder4['mpg'], ddof=1) # Calculate the sample standard deviation
n = len(cylinder4['mpg']) # Calculate n for the sample

alpha = 0.05 # Set significance level
t_crit = t.ppf(1-alpha/2, n-1) # Calculate critical t-value
sem = std / np.sqrt(n) # Calculate the standard error of the mean 

upper = mean - t_crit * sem # Calculate below limit
lower = mean + t_crit * sem # Calculate upper limit
round(upper, 1), round(lower, 1) # Rounding to 1 decimal place
print('Result: ' + str(round(mean,1)) + ' \u00B1 ' + str(round(t_crit*sem,1))) # print confidenceinterval
print(f'Confidence interval (\u03B1=0.05) for average fuel efficiency of cars with 4 cylinders : [{round(upper,1)} {round(lower,1)}]') # Printa konfidensintervall


Result: 29.3 ± 0.8
Confidence interval (α=0.05) for average fuel efficiency of cars with 4 cylinders : [28.5 30.1]


So 95% confidence interval for average fuel efficiency of cars with 4 cylinders would be [28.5 30.1]

#### what is the 90% confidence interval for the difference in horsepower between cars with 4 cylinders and cars with 8 cylinders?


If the populations standard deviation is equal then konfidence interval for $\mu_{1}$- $\mu_{2}$  

$(\bar{x_1} - \bar{x_2})\pm t_{\alpha/2, {df}}{\sqrt{\frac{s_p^2}{n_1} + \frac{s_p^2}{n_2}}}$

$s_{p} = {\sqrt{\frac{({n_1}-1){s_1^2} + ({n_2}-1){s_2^2}}{({n_1}+{n_2}-2)}}}  $

If the populations standard deviation is not equal then konfidence interval for $\mu_{1}$- $\mu_{2}$  

$(\bar{x_1} - \bar{x_2})\pm t_{\alpha/2, {df}}{\sqrt{\frac{s_1^2}{n_1} + \frac{s_2^2}{n_2}}}$




Assuming tha data of the population standard deviations are equal

In [48]:
cylinder_4 = cars[cars['cylinders'] == 4]
cylinder_8 = cars[cars['cylinders'] == 8]

n_1 = len(cylinder_4) #  Calculate n for the sample of cars with 4 cylinders 
n_2 = len(cylinder_8) # Calculate n for the sample of cars with 8 cylinders 

x_1 = np.mean(cylinder_4['horsepower']) # Calculate the sample 1 mean
s_1 = np.std(cylinder4['horsepower'], ddof=1) # Calculate the sample 1 standard deviation

x_2 = np.mean(cylinder_8['horsepower']) # Calculate the sample 2 mean
s_2 = np.std(cylinder_8['horsepower'], ddof=1) # Calculate the sample 2 standard deviation

alpha = 0.1 # Set significance level
df = n_1+n_2-2 # degree of fredom
t_crit = t.ppf(1-alpha/2, df ) # Calculate critical t-value

s_p = np.sqrt(((n_1-1)*s_1**2 + (n_2-1)*s_2**2)/(n_1+n_2-2)) # calculate standard deviation of two samples
sem = s_p * np.sqrt((1/n_1)+(1/n_2)) # Calculate the standard error of the mean

diff_mean = x_1-x_2 # calculate difference in mean
upper = diff_mean - t_crit * sem # Calculate below limit
lower = diff_mean + t_crit * sem # Calculate upper limit

print('Result: ' + str(round(diff_mean,1)) + ' \u00B1 ' + str(round(t_crit*sem,1))) # print confidenceinterval
print(f'Confidence interval (\u03B1=0.1) for the difference in horsepower between cars with 4 cylinders and cars with 8 cylinders : [{round(upper,1)} {round(lower,1)}]') # Printa konfidensintervall




Result: -80.0 ± 4.1
Confidence interval (α=0.1) for the difference in horsepower between cars with 4 cylinders and cars with 8 cylinders : [-84.1 -75.9]


So 90% confidence interval for the difference in horsepower between cars with 4 cylinders and cars with 8 cylinders would be [-84.1 -75.9]